src/v/kafka/data/cloud_topic_partition.cc

```diff
@@ -0,0 +1,599 @@
+        return error_code::offset_not_available;
+    }
+    auto ot_state = _partition->get_offset_translator_state();
+    return ot_state->from_log_offset(maybe_lso);
+}
+
+bool cloud_topic_partition::is_leader() const {
+    return _partition->is_leader();
+}
+
+ss::future<std::error_code> cloud_topic_partition::linearizable_barrier() {
+    auto r = co_await _partition->linearizable_barrier();
+    if (r) {
+        co_return raft::errc::success;
+    }
+    co_return r.error();
+}
+
+cluster::partition_probe& cloud_topic_partition::probe() {
+    return _partition->probe();
+}
+
+kafka::leader_epoch cloud_topic_partition::leader_epoch() const {
+    return leader_epoch_from_term(_partition->raft()->confirmed_term());
+}
+
+ss::future<storage::translating_reader> cloud_topic_partition::make_reader(
+  storage::log_reader_config cfg,
+  std::optional<model::timeout_clock::time_point> timeout) {
+    vassert(_ct_api != nullptr, "cloud topics api not initialized");
+
+    auto ot_state = _partition->get_offset_translator_state();
+
+    cfg.start_offset = ot_state->to_log_offset(cfg.start_offset);
+    cfg.max_offset = ot_state->to_log_offset(cfg.max_offset);
+    cfg.translate_offsets = storage::translate_offsets::yes;
+    cfg.type_filter = {model::record_batch_type::dl_placeholder};
+
+    // TODO: add code path that fetches metadata from the dl_stm for L1 read
+    // path
+
+    // This part comprises the control plane query. The partition is a metadata
+    // storage for the cloud topic.
+    auto underlying = co_await _partition->make_reader(cfg, timeout);
+    auto placeholders = co_await model::consume_reader_to_chunked_vector(
+      std::move(underlying), model::no_timeout);
+
+    // TODO: convert to a extent_meta
+    auto extents = [ph = std::move(placeholders)]() mutable {
+        chunked_vector<experimental::cloud_topics::extent_meta> res;
+        for (auto&& batch : ph) {
+            experimental::cloud_topics::extent_meta e{
+              .base_offset = model::offset_cast(batch.base_offset()),
+              .last_offset = model::offset_cast(batch.last_offset()),
+            };
+            iobuf payload = std::move(batch).release_data();
+            iobuf_parser parser(std::move(payload));
+            auto record = model::parse_one_record_from_buffer(parser);
+            iobuf value = std::move(record).release_value();
+            auto placeholder
+              = serde::from_iobuf<experimental::cloud_topics::dl_placeholder>(
+                std::move(value));
+            e.id = placeholder.id;
+            e.first_byte_offset = placeholder.offset;
+            e.byte_range_size = placeholder.size_bytes;
+            res.push_back(e);
+        }
+        return res;
+    }();
+
+    // This part comprises the data plane query. The 'cloud_topics::api' is
+    // responsible for moving the data from the cloud storage to the local
+    // cache.
+    auto data_batches = co_await _ct_api->materialize(
```

Comment:
This is fine for now. Longer term should this return a reader so it can incrementally make progress. If there are multiple fetches then we currently have to wait for all of them before returning more infomation.